docs/docs/unit-testing.md

```diff
@@ -0,0 +1,257 @@
+---
+title: Unit testing
+---
+
+Unit testing is a great way to protect against errors in your code before you
+deploy it. While Gatsby does not include support for unit testing out of the
+box, it only takes a few steps to get up and running. However there are a few
+features of the Gatsby build process that mean the standard Jest setup doesn't
+quite work. This guide shows you how to set it up.
+
+## Setting up your environment
+
+The most popular testing framework for React is [Jest](https://jestjs.io/),
+which was created by Facebook. While Jest is a general purpose JavaScript unit
+testing framework, it has lots of features that make it work particularly well
+with React.
+
+For this guide, you will be starting with `gatsby-starter-blog`, but the
+concepts should be the same or very similar for your site.
+
+First you need to install Jest and some more required packages. You need to
+install Babel 7 as it's required by Jest.
+
+```sh
+npm install --save-dev jest babel-jest react-test-renderer identity-obj-proxy 'babel-core@^7.0.0-0' @babel/core @babel/preset-env @babel/preset-react @babel/plugin-proposal-class-properties @babel/plugin-proposal-optional-chaining
+```
+
+Because Gatsby handles its own Babel configuration, you will need to manually
+tell Jest to use `babel-jest`. The easiest way to do this is to add a `"jest"`
+section in your `package.json`. You can set up some useful defaults at the same
+time:
+
+```json
+  "jest": {
+    "transform": {
+      "^.+\\.jsx?$": "<rootDir>/jest-preprocess.js"
+    },
+    "testRegex": "/.*(__tests__\\/.*)|(.*(test|spec))\\.jsx?$",
+    "moduleNameMapper": {
+      ".+\\.(css|styl|less|sass|scss)$": "identity-obj-proxy",
+      ".+\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.js"
+    },
+    "testPathIgnorePatterns": ["node_modules", ".cache"],
+    "transformIgnorePatterns": [
+      "node_modules/(?!(gatsby)/)"
+    ]
+  }
+```
+
+The `transform` section tells Jest that all `js` or `jsx` files need to be
+transformed using a `jest-preprocess.js` file in the project root. Go ahead and
+create this file now. This is where you set up your Babel config. You can start
+with a minimal config.
+
+```js
+// jest-preprocess.js
+const babelOptions = {
+  presets: ["@babel/react", "@babel/env"],
+  plugins: [
+    "@babel/plugin-proposal-optional-chaining",
+    "@babel/plugin-proposal-class-properties",
+  ],
+}
+
+module.exports = require("babel-jest").createTransformer(babelOptions)
+```
+
+Back to the Jest config, you can see the next option is `testRegex`. This is the
+pattern telling Jest which files contain tests. The pattern above matches any
+`.js` file inside a `__tests__` directory, or any file elsewhere with the
+extension `.test.js` or `.spec.js`. You are telling Jest to ignore any tests in
+the `node_modules` or `.cache` directories.
+
+The `moduleNameMapper` section works a bit like webpack rules, and tells Jest
```

Comment:
Webpack

Reply:
I just changed this in my last review :) webpack is officially written as 'webpack' rather than 'Webpack'. Awkward, I know! See usage on https://webpack.js.org/

Reply:
oh wut. TIL. Thanks!